File: code/C++/31-Next Permutation.cpp
/*
Next Permutation
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

The replacement must be in-place, do not allocate extra memory.

Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
1,2,3 → 1,3,2
3,2,1 → 1,2,3
1,1,5 → 1,5,1

他喵的早这么说不就可以了么

http://www.cnblogs.com/dragonpig/archive/2010/01/21/1653680.html

http://fisherlei.blogspot.com/2012/12/leetcode-next-permutation.html
*/

class Solution {
public:
    void nextPermutation(vector<int> &num) {
        int i,j,len=num.size();
	//bool flg=true;
	
	for (i=len-2;i>=0;i--) {
		if(num[i+1]>num[i]){
		//	flg=false;
			for (j=len-1;j>=i;j--) {
				if(num[j]>num[i])
				  break;
			}
			swap(num[i],num[j]);
			reverse(num.begin()+i+1,num.end());
			return ;
		}
    }
    reverse(num.begin(),num.end());
	return ;
    }
};

File: code/Python/31-Next Permutation.py
# -*- coding: utf-8 -*
"""
@author: Kai Chen
@file: 31-Next Permutation.py
@time: 2020/11/10 9:23
@desc:
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such an arrangement is not possible, it must rearrange it as the lowest possible order (i.e., sorted in ascending order).

The replacement must be in place and use only constant extra memory.



Example 1:

Input: nums = [1,2,3]
Output: [1,3,2]
Example 2:

Input: nums = [3,2,1]
Output: [1,2,3]
Example 3:

Input: nums = [1,1,5]
Output: [1,5,1]
Example 4:

Input: nums = [1]
Output: [1]


Constraints:

1 <= nums.length <= 100
0 <= nums[i] <= 100
"""


class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        for i in range(n - 2, -1, -1):
            if nums[i + 1] > nums[i]:
                for j in range(n - 1, i - 1, -1):
                    if nums[j] > nums[i]:
                        break
                nums[i], nums[j] = nums[j], nums[i]
                nums[i+1:] = list(reversed(nums[i+1:]))
                return
        nums[:] = list(reversed(nums[:]))
        return

